2020-04-25-IJCTF/klepto/README.md

```diff
@@ -0,0 +1,95 @@
+# Klepto (crypto, 932p, 12 solved)
+
+In the task we get some [code](klepto.py) and [results](enc).
+
+Result is just RSA public key and ciphertext.
+The goal is to analyse the key generation logic and break the public key.
+
+## Solution
+
+The generation code might look complex, but most of it is just primarity-check, so we can skip that entirely.
+The only useful part is:
+
+```python
+IV = 5326453656607417485924926839199132413931736025515611536784196637666048763966950827721156230379191350953055349475955277628710716247141676818146987326676993279104387449130750616051134331541820233198166403132059774303658296195227464112071213601114885150668492425205790070658813071773332779327555516353982732641; seed = 0; temp = [0, 0]; key = 0
+while(key != 2):
+    if key == 0:
+        seed = getrandbits(1024) | (2 ** 1023 + 1)
+    seed_ = seed ^ IV
+    n = seed_ << 1024 | getrandbits(1024)
+    seed = n//seed | 1
+    while(1):
+        seed += 2;
+        ## primarity checks
+    temp[key] = seed
+return(temp[0], temp[1])
+```
+
+Let's look closely at:
+
+```python
+seed_ = seed ^ IV
+n = seed_ << 1024 | getrandbits(1024)
+seed = n//seed | 1
+```
+
+Let's rewrite this:
+
+```
+s2 = seed ^ IV
+n = s2 * 2**1024 + r
+almost_p = n//seed
+p = almost_p + small_value
+```
+
+Now:
+
+```
+almost_p = n//seed
+almost_p = s2 * 2**1024 + r//seed
```

Comment:
Shouldn't this be `almost_p = (s2 * 2**1024 + r) // seed` ?

Reply:
Fixed, I also added a few more parenthesis to avoid confusion with the order of XOR and division/multiplication.